tests/test_protocols/test_generator/test_extract_specification.py

```diff
@@ -0,0 +1,533 @@
+            "content_dict_int_int": "Dict[int, int]",
+            "content_dict_int_float": "Dict[int, float]",
+            "content_dict_int_bool": "Dict[int, bool]",
+            "content_dict_int_str": "Dict[int, str]",
+            "content_dict_bool_bytes": "Dict[bool, bytes]",
+            "content_dict_bool_int": "Dict[bool, int]",
+            "content_dict_bool_float": "Dict[bool, float]",
+            "content_dict_bool_bool": "Dict[bool, bool]",
+            "content_dict_bool_str": "Dict[bool, str]",
+            "content_dict_str_bytes": "Dict[str, bytes]",
+            "content_dict_str_int": "Dict[str, int]",
+            "content_dict_str_float": "Dict[str, float]",
+            "content_dict_str_bool": "Dict[str, bool]",
+            "content_dict_str_str": "Dict[str, str]",
+            "content_union_1": "Union[DataModel, bytes, int, float, bool, str, FrozenSet[int], Tuple[bool, ...], Dict[str, int]]",
+            "content_union_2": "Union[FrozenSet[bytes], FrozenSet[int], FrozenSet[str], Tuple[float, ...], Tuple[bool, ...], Tuple[bytes, ...], Dict[str, int], Dict[int, float], Dict[bool, bytes]]",
+            "content_o_ct": "Optional[DataModel]",
+            "content_o_bool": "Optional[bool]",
+            "content_o_set_int": "Optional[FrozenSet[int]]",
+            "content_o_list_bytes": "Optional[Tuple[bytes, ...]]",
+            "content_o_dict_str_int": "Optional[Dict[str, int]]",
+        }
+        assert spec.all_custom_types == ["DataModel"]
+        assert spec.custom_custom_types == {"DataModel": "CustomDataModel"}
+        assert spec.initial_performatives == ["PERFORMATIVE_CT", "PERFORMATIVE_PT"]
+        assert spec.reply == {
+            "performative_ct": ["performative_pct"],
+            "performative_pt": ["performative_pmt"],
+            "performative_pct": ["performative_mt", "performative_o"],
+            "performative_pmt": ["performative_mt", "performative_o"],
+            "performative_mt": [],
+            "performative_o": [],
+            "performative_empty_contents": ["performative_empty_contents"],
+        }
+        assert spec.terminal_performatives == [
+            "PERFORMATIVE_MT",
+            "PERFORMATIVE_O",
+            "PERFORMATIVE_EMPTY_CONTENTS",
+        ]
+        assert spec.roles == ["role_1", "role_2"]
+        assert spec.end_states == ["end_state_1", "end_state_2", "end_state_3"]
+        assert spec.typing_imports == {
+            "Set": True,
+            "Tuple": True,
+            "cast": True,
+            "FrozenSet": True,
+            "Dict": True,
+            "Union": True,
+            "Optional": True,
+        }
+
+    @mock.patch(
+        "aea.protocols.generator.validate.validate",
+        return_value=[False, "some error."],
+    )
+    def test_extract_negative_invalid_specification(self, validate_mock):
+        """Negative test the 'extract' method."""
+        pytest.skip("todo")
+        # ToDo
```

Comment:
TODO should always have a comment what's todo

Reply:
`# ToDo: Complete this test!`